compiler/rustc_hir_typeck/src/pat.rs

```diff
@@ -562,6 +577,9 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                     type between two end-points, you can use a guard.",
             );
         }
+        if is_ty_references_error {
```

Comment:
I wonder if we can just do `(lhs, rhs).references_error()` here? I think 3-tuples implement `TypeVisitable`... or else they should.

Reply:
The tuple they've got is `(Option(_, ty, _), Option(_, ty, _))`, they wouldn't be able to invoke it in any case without changing the logic. I do not remember if `delay_as_bug` is idempotent. If it is, then we could pull up line 581 into every place above where `is_ty_references_error` is assigned instead.

Reply:
[`downgrade_to_delayed_bug`](https://doc.rust-lang.org/nightly/nightly-rustc/src/rustc_errors/diagnostic_builder.rs.html#489-492) is idempotent

Reply:
I used [downgrade_to_delayed_bug](https://doc.rust-lang.org/nightly/nightly-rustc/src/rustc_errors/diagnostic_builder.rs.html#489-492) to replace `delay_as_bug()`.
since` (lhs, rhs).references_error()` works, we don't need to pull up `is_ty_references_error `which was removed.